Auth0.Net/Client.cs

```diff
@@ -320,17 +320,29 @@ public TokenResult ExchangeAuthorizationCodePerAccessToken(string code, string r
             request.AddParameter("grant_type", "authorization_code", ParameterType.GetOrPost);
             request.AddParameter("redirect_uri", redirectUri, ParameterType.GetOrPost);
 
-            var response = this.client.Execute<Dictionary<string, string>>(request).Data;
+            var response = this.client.Execute<Dictionary<string, string>>(request);
+            var body = response.Data;
 
-            if (response.ContainsKey("error") || response.ContainsKey("error_description"))
+            if (response.ResponseStatus != ResponseStatus.Completed)
             {
-                throw new OAuthException(response["error_description"], response["error"]);
+                throw new OAuthException(
+                    "The Auth0 API did not return a complete response; ResponseStatus=" + response.ResponseStatus + "; ErrorMessage=" + response.ErrorMessage ?? string.Empty, 
+                    string.Empty);
+            }
+
+            if (response.StatusCode != HttpStatusCode.OK)
+            {
+                var errorDescription = body.ContainsKey("error_description") ? body["error_description"] : "(no description)";
+                errorDescription += "; StatusCode=" + (int)response.StatusCode;
+                var errorCode = body.ContainsKey("error") ? body["error"] : string.Empty;
+
+                throw new OAuthException(errorDescription, errorCode);
             }
 

```

Comment:
It might be worth checking `response.Data` here. Not sure if it can be null or not, but if it can it would be better to throw here than when `body.ContainsKey` fails.

```cs
var body = response.Data;
if (body == null) { 
  throw AppropriateExcecption('blah blah');
}
```